### PR TITLE
Extract `on` transitions

### PR DESCRIPTION
**Pull request by Andarist**

### CODE REVIEW

File: new-packages/ts-project/__tests__/transitions.test.ts [hunk 1]
```diff
@@ -0,0 +1,2235 @@
...
+  const tmpPath = await testdir({
+    'tsconfig.json': JSON.stringify({}),
+    'index.ts': ts`
+      import { createMachine } from "xstate";
+
+      createMachine({
+        states: {
+          foo: {
+            on: {
+              FOO: { target: "bar" },
```
Comment (farskid) [inline new-packages/ts-project/__tests__/transitions.test.ts:350]: This example doesn't match test title.

File: new-packages/ts-project/__tests__/transitions.test.ts [hunk 2]
```diff
@@ -0,0 +1,2235 @@
...
+test('should extract transition with multiple targets (array of strings)', async () => {
+  const tmpPath = await testdir({
+    'tsconfig.json': JSON.stringify({}),
+    'index.ts': ts`
+      import { createMachine } from "xstate";
+
+      createMachine({
+        on: {
+          MULTIPLE: {
+            target: [".a.a2", ".b.b2"],
```
Comment (farskid) [inline new-packages/ts-project/__tests__/transitions.test.ts:1130]: nit: This test can be divided into two tests. One testing simple target strings in array and another testing deep child `.a.b.c`.

File: new-packages/ts-project/__tests__/transitions.test.ts [hunk 3]
```diff
@@ -0,0 +1,2235 @@
...
+                    "eventType": "MULTIPLE",
+                    "type": "named",
+                  },
+                  "guard": undefined,
+                  "internal": true,
+                },
+                "source": "state-0",
+                "targets": [
+                  "state-3",
+                  "state-6",
```
Comment (farskid) [inline new-packages/ts-project/__tests__/transitions.test.ts:1179]: Why do we have both of these states as the target?

File: new-packages/ts-project/__tests__/transitions.test.ts [hunk 4]
```diff
@@ -0,0 +1,2235 @@
...
+            },
+            "root": "state-0",
+          },
+          [],
+        ],
+      ]
+    `);
+});
+
+test('should extract forbidden transition (null)', async () => {
```
Comment (farskid) [inline new-packages/ts-project/__tests__/transitions.test.ts:1405]: This test is duplicated.

File: new-packages/ts-project/__tests__/transitions.test.ts [hunk 5]
```diff
@@ -0,0 +1,2235 @@
...
+            },
+            "root": "state-0",
+          },
+          [],
+        ],
+      ]
+    `);
+});
+
+test('should extract transition description', async () => {
```
Comment (farskid) [inline new-packages/ts-project/__tests__/transitions.test.ts:2141]: nit: if we're going to have a single test for description, I'd rather have it be a multiline description. We've had problems with those before.

**Comment by farskid** [comment]
@Andarist Are we not extracting always and after transitions in this PR?

**Comment by Andarist** [comment]
> @Andarist Are we not extracting always and after transitions in this PR?

I'm working on `always` right now. I plan to just add those in followup PRs to make each PR smaller and easier for review etc.

File: new-packages/ts-project/__tests__/transitions.test.ts (hunk 3 above)
Comment (Andarist) [reply new-packages/ts-project/__tests__/transitions.test.ts:1179]: The test title is:
> should extract transition with multiple targets (array of strings)

Those are two parallel regions that are targeted here by a single transition

File: new-packages/ts-project/__tests__/transitions.test.ts (hunk 5 above)
Comment (Andarist) [reply new-packages/ts-project/__tests__/transitions.test.ts:2141]: added a multiline one now

File: new-packages/ts-project/__tests__/transitions.test.ts (hunk 1 above)
Comment (Andarist) [reply new-packages/ts-project/__tests__/transitions.test.ts:350]: fixed

File: new-packages/ts-project/__tests__/transitions.test.ts (hunk 3 above)
Comment (farskid) [reply new-packages/ts-project/__tests__/transitions.test.ts:1179]: Is that how Studio behaves too?

Comment (farskid) [reply new-packages/ts-project/__tests__/transitions.test.ts:1179]: Quickly tested and this is not an edge Studio can handle. Honestly, multi target is a new concept to me.

**Review by farskid — APPROVED** [review]

File: new-packages/ts-project/src/state.ts [hunk 6]
```diff
@@ -51,12 +63,72 @@ export function createActorBlock({
...
+  };
+}
+
+function isForbiddenTarget(
+  ts: typeof import('typescript'),
+  element: Expression,
+): boolean {
+  return (
+    isUndefined(ts, element) ||
+    // null isn't technically allowe by the XState's types but it behaves in the same way at runtime
```
Comment (farskid) [inline new-packages/ts-project/src/state.ts:97]:
```suggestion
    // null isn't technically allowed by the XState's types but it behaves in the same way at runtime
```

File: new-packages/ts-project/__tests__/transitions.test.ts (hunk 3 above)
Comment (Andarist) [reply new-packages/ts-project/__tests__/transitions.test.ts:1179]: That's something to be fixed on the Studio's side.

**Pull request merged**